pkg/oauthserver/authenticator/password/keystonepassword/keystonepassword.go

```diff
@@ -4,36 +4,105 @@ import (
 	"fmt"
 	"net/http"
 	"runtime/debug"
+	"encoding/json"
 
 	"github.com/golang/glog"
+	authapi "github.com/openshift/origin/pkg/oauthserver/api"
 	"github.com/rackspace/gophercloud"
 	"github.com/rackspace/gophercloud/openstack"
+	tokens3 "github.com/rackspace/gophercloud/openstack/identity/v3/tokens"
 
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/apiserver/pkg/authentication/authenticator"
 	"k8s.io/apiserver/pkg/authentication/user"
-
-	authapi "github.com/openshift/origin/pkg/oauthserver/api"
-)
+	)
 
 // keystonePasswordAuthenticator uses OpenStack keystone to authenticate a user by password
 type keystonePasswordAuthenticator struct {
-	providerName   string
-	url            string
-	client         *http.Client
-	domainName     string
-	identityMapper authapi.UserIdentityMapper
+	providerName           string
+	url                    string
+	client                 *http.Client
+	domainName             string
+	identityMapper         authapi.UserIdentityMapper
+	useKeystoneIdentity    bool
 }
 
 // New creates a new password authenticator that uses OpenStack keystone to authenticate a user by password
 // A custom transport can be provided (typically to customize TLS options like trusted roots or present a client certificate).
 // If no transport is provided, http.DefaultTransport is used
-func New(providerName string, url string, transport http.RoundTripper, domainName string, identityMapper authapi.UserIdentityMapper) authenticator.Password {
+func New(providerName string, url string, transport http.RoundTripper, domainName string, identityMapper authapi.UserIdentityMapper, useKeystoneIdentity bool) authenticator.Password {
 	if transport == nil {
 		transport = http.DefaultTransport
 	}
 	client := &http.Client{Transport: transport}
-	return &keystonePasswordAuthenticator{providerName, url, client, domainName, identityMapper}
+	return &keystonePasswordAuthenticator{providerName, url, client, domainName, identityMapper, useKeystoneIdentity}
+}
+
+// Authenticate user and return his ID from Keystone
+func GetUserIDv3(client *gophercloud.ProviderClient, options gophercloud.AuthOptions) (string, error) {
+	// Override the generated service endpoint with the one returned by the version endpoint.
+	v3Client := openstack.NewIdentityV3(client)
+
+	// copy the auth options to a local variable that we can change. `options`
+	// needs to stay as-is for reauth purposes
+	v3Options := options
+
+	var scope *tokens3.Scope
+	if options.TenantID != "" {
+		scope = &tokens3.Scope{
+			ProjectID: options.TenantID,
+		}
+		v3Options.TenantID = ""
+		v3Options.TenantName = ""
+	} else {
+		if options.TenantName != "" {
+			scope = &tokens3.Scope{
+				ProjectName: options.TenantName,
+				DomainID:    options.DomainID,
+				DomainName:  options.DomainName,
+			}
+			v3Options.TenantName = ""
+		}
+	}
+
+	result := tokens3.Create(v3Client, v3Options, scope)
```

Comment:
is this always creating a token, even in cases where we don't need it? does that need to be cleaned up?

Reply:
This is how we authenticate in keystone. If an authentication was successful, keystone returns a token inside the header, and full information about the user inside the response body, otherwise an error.
https://developer.openstack.org/api-ref/identity/v3/#password-authentication-with-unscoped-authorization

Can you give me an example of the cases that you are talking about?